Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/PredictorCorrector.java

```diff
@@ -250,45 +266,53 @@ protected boolean trigger(DiVector candidate, int gap, int baseDimension, DiVect
+                || (method == TransformMethod.NORMALIZE_DIFFERENCE);
+        double[] scale = result.getScale();
+        double[] shift = result.getShift();
+        // backward compatibility of older models
         if (scale == null || shift == null) {
             return true;
         }
         boolean answer = false;
-        for (int y = 0; y < baseDimensions && !answer; y++) {
-            double scaleFactor = (scale == null) ? 1.0 : scale[y];
-            double delta = Math.abs(point[startPosition + y] - newPoint[startPosition + y]) * scaleFactor;
-            double shiftBase = (shift == null) ? 0 : shift[y];
-            double shiftAmount = 0;
-            if (scaleFactor != 1.0 || shiftBase != 0) {
-                shiftAmount += 2 * DEFAULT_NORMALIZATION_PRECISION * Math.abs(shiftBase);
-            }
+        // only for input dimensions, for which scale is defined currently
+        for (int y = 0; y < result.getInputLength() && !answer; y++) {
+            double observedGap = Math.abs(point[startPosition + y] - newPoint[startPosition + y]);
+            double pathGap = calculatePathDeviation(point, startPosition, y, baseDimensions, differenced);
+            if (observedGap > 0.1 * pathGap) {
+                double scaleFactor = (scale == null) ? 1.0 : scale[y];
+                double delta = observedGap * scaleFactor;
+                double shiftBase = (shift == null) ? 0 : shift[y];
+                double shiftAmount = 0;
+                if (scaleFactor != 1.0 || shiftBase != 0) {
+                    double multiplier = (method == TransformMethod.NORMALIZE) ? 4 : 2;
+                    shiftAmount += multiplier * DEFAULT_NORMALIZATION_PRECISION * Math.abs(shiftBase);
+                }
 
-            // note that values cannot be reconstructed well if differencing was invoked
-            double a = Math.abs(scaleFactor * point[startPosition + y] + shiftBase);
-            double b = Math.abs(scaleFactor * newPoint[startPosition + y] + shiftBase);
+                // note that values cannot be reconstructed well if differencing was invoked
+                double a = Math.abs(scaleFactor * point[startPosition + y] + shiftBase);
+                double b = Math.abs(scaleFactor * newPoint[startPosition + y] + shiftBase);
 
-            // for non-trivial transformations
-            if (scaleFactor != 1.0 || shiftBase != 0) {
-                shiftAmount += DEFAULT_NORMALIZATION_PRECISION * (scaleFactor + (a + b) / 2);
-            }
-            answer = significantScore || (delta > shiftAmount + DEFAULT_NORMALIZATION_PRECISION);
-            if (answer) {
-                answer = (a < b - ignoreNearExpectedFromBelow[y]) || (a > b + ignoreNearExpectedFromAbove[y]);
+                // for non-trivial transformations
+                if (scaleFactor != 1.0 || shiftBase != 0) {
+                    double multiplier = (method == TransformMethod.NORMALIZE) ? 2 : 1;
```

Comment:
Great issue to raise. Yes - Normalized_Difference and Difference can be small. They actually point to the factor 2 -- because of the differencing operation, when we recreate the number in the original input space -- it seems that they need 2 additions (one for shift and one for previous number). So the condition number is at least 2 for the addition part. Maybe for SUBTRACT_MA (moving average) we would have use a condition number 1, but it risked the code being even more complicated. 

In contrast, the scaling operation is performed once - loss of precision is lower for multiplication. The second block of 2 vs 1 also corresponds to relative error -- which would be 0.1% . 

Normalization, by virtue of squeezing all coordinates towards 0 does generate more errors. It can be 2X and X for values of other than 2; but this setting seemed to make the results comparable across the difference transformations for the same fixed input. These constants can be tuned more.